### ISSUE TITLE
sriov network plugin does not work with kubevirt 0.24.0

### ISSUE THREAD

**Issue by anandgbhat** [kubevirt/kubevirt#3035]
/kind bug

**What happened**:

VM launch with kubevirt version 0.24.0 fails with the following error:

{"component":"virt-launcher","kind":"","level":"error","msg":"Failed to sync vmi","name":"vmi-sriov","namespace":"default","pos":"server.go:161","reason":"virError(Code=67, Domain=10, Message='unsupported configuration: host doesn't support passthrough of host PCI devices')","timestamp":"2019-12-18T16:03:45.615936Z","uid":"e0cdbd24-f65b-4c0d-b0cb-9f2cc9995ba7"}

**What you expected to happen**:
It should have worked !!

**How to reproduce it (as minimally and precisely as possible)**:
Create SRIO-V interfaces and try to launch a VM with SRIO-V interfaces. VM launch fails with the above error.

**Anything else we need to know?**:
VM launch with kubevirt version 0.24.0 fails with the following error:

{"component":"virt-launcher","kind":"","level":"error","msg":"Failed to sync vmi","name":"vmi-sriov","namespace":"default","pos":"server.go:161","reason":"virError(Code=67, Domain=10, Message='unsupported configuration: host doesn't support passthrough of host PCI devices')","timestamp":"2019-12-18T16:03:45.615936Z","uid":"e0cdbd24-f65b-4c0d-b0cb-9f2cc9995ba7"}

Inside the virt-launcher pod, virt-host-validate fails with following error:
QEMU: Checking for device assignment IOMMU support : WARN (No ACPI DMAR table found, IOMMU either disabled in BIOS or not supported by this hardware platform)

This is appears to be of because /sys/firmware/acpi not available inside the virt-launcher container without privileged container. It appears kubervirt 0.24.0 launches virt-launcher pod without privileges and does not mount /sys/firmware/acpi.
Updating the code run launcher as a privileged container fixed this issue. But this resulted in other errors including "Failed to unbind PCI device '0000:xx:x.x' from ixgbevf". Remounting /sys as as rw fixed this issue but /dev/vfio/xxx wasn't found inside the container.

It appears container not running in privilege mode is causing this issue. This is default in kubevirt 0.24.0 and was done as a security measure. This breaks srio-cni . This needs a fix.

**Environment**:
- KubeVirt version (use `virtctl version`): 0.24
- Kubernetes version (use `kubectl version`): 1.10
- VM or VMI specifications: 
- Cloud provider or hardware configuration: 
- OS (e.g. from /etc/os-release):  CentOS 7.7
- Kernel (e.g. `uname -a`): 3.10.0-1062
- Install tools:
- Others:

**Comment by rmohr** [comment]
@anandgbhat that is interesting. We have e2e tests with sriov cards, so at least the sriov device plugins work for intel and melanox. I can see that there exist different sriov cni plugins. Which one are you using?

@phoracek maybe you can have a look?

**Comment by phoracek** [comment]
@anandgbhat I double Roman. Could you please describe how did you install SR-IOV plugin? Our CI is not running with the latest one, so there may have been an issue introduced.

**Comment by rmohr** [comment]
@anandgbhat could you also share the logs from the virt-launcher container? There are also some bogus errors, where libvirt simply logs that specific features are not present, but which don't have an impact on kubevirt.

**Comment by anandgbhat** [comment]
We used Intel SRIO-V plugin with SRIO-V device plugin to manage VFs.

https://github.com/intel?utf8=%E2%9C%93&q=sriov&type=&language=

**Comment by phoracek** [comment]
I will bump the version of the SR-IOV operator we use on our CI to see if it reproduces https://github.com/kubevirt/kubevirtci/pull/277.

@anandgbhat could you share what is the vendor of the NIC you use? Also, it would be helpful if you shared the `NetworkAttachmentDefinition` you have for the SR-IOV connection.

I would advice you to try https://github.com/openshift/sriov-network-operator. It supports Kubernetes too and it would do a lot of host configuration for you.

**Comment by booxter** [comment]
@anandgbhat I assume you deploy SR-IOV components (CNI, DP) manually, not via sriov-network-operator that Petr linked to. I would highly advise you to try the operator in your environment, it will automate both deploying components into cluster as well as configuring kernel (even rebooting the machine when needed to apply the new command line arguments). Regardless...

The original error message from libvirt suggests that you don't have IOMMU enabled on the machine. This may be either a BIOS configuration issue or kernel needs additional command line arguments to enable it on boot.

For the first suggestion - BIOS - refer to your hardware documentation to see how to enable IOMMU on the machine. For the second suggestion - kernel - please make sure you have the following argument passed to the kernel on boot: `intel_iommu=on`. If it doesn't help, consider also adding `iommu=pt`. If it still doesn't work, please check `dmesg` output to see if there are additional kernel warnings specific to your platform. (Some older hardware may require additional quirks enabled to make SR-IOV passthrough work.)

Regardless, please share kernel options you use to boot the machine. You can extract them from `/proc/cmdline`.

As for your suggestion that `kubevirt` may need to run `launcher` pod as privileged or with additional volumes mounted into it, I don't think so. I am not exactly sure how `firmware` volume is related to your particular issue though. Perhaps you could explain how you got to the idea to try adding it. The initial description doesn't clarify it.

Long story short, 1) please try with sriov-network-operator if possible; 2) please share kernel arguments used to boot the machine; 3) please check that IOMMU is enabled in BIOS; 4) if IOMMU arguments are not used to boot the kernel, please try the suggestions above (this requires machine reboot; BIOS obviously too).

Thanks.

**Comment by anandgbhat** [comment]
@booxter Thanks for looking at this. We will try using sriov operator. 

Between, we could circumvent the issue using some hacks. IOMMU was enabled using kernel command line and VT-D was also enabled. Issue was due to virt-launcher not being privileged container. Same setup worked well with version 0.17 which had virt-launcher container as a privileged container.

I understand that privileged containers are security risks and disabling it is the correct way forward. But this broke SRIO-V functionality.

**Comment by booxter** [comment]
@anandgbhat it did not break SR-IOV per se - we have it working with sriov-network-operator in gate CI. Every patch in gate is tested against the SR-IOV job. So whatever the issue you have is probably some form of configuration issue, either BIOS or kernel or incorrect versions of SR-IOV plugins. Now, I cannot guarantee there is no obscure bug in what you describe, but definitely SR-IOV works in configuration we test against in gate. You shouldn't require making the virt-launcher privileged to make it work, so exploring this path is not a long term solution.

Please share the exact kernel arguments you passed to enable IOMMU, and please clarify how you enabled the feature in your BIOS. Please also share kernel message log (`dmesg`) so that we can check that you have IOMMU enabled and working.

**Comment by booxter** [comment]
Another question I have is: you said in the description:

> Remounting /sys as as rw fixed this issue but /dev/vfio/xxx wasn't found inside the container.

Considering this, and considering that we don't support environments that don't have their VFs pre-registered with VFIO (in `sriov-network-operator` enabled environment, the operator takes care of picking the right kernel driver for each VF, in case of kubevirt it's `vfio-pci`), could you please check that your deployment / host configuration procedure includes registering all VFs exposed via SR-IOV device plugin to the cluster with VFIO? Let us know how you make sure the registration happens on host boot / cluster deployment.

Thanks.

**Comment by booxter** [comment]
Please also check that `/etc/pcidp/config.json` defines a resource pool of `vfio` type and not `netdevice`. The latter was once used in earlier versions of `kubevirt` but is no longer supported. Instead, VFs are pre-registered with VFIO, device plugin is configured for `vfio` which makes it allocate the corresponding `/dev/vfio/{DEVICE}` to the `virt-launcher` container, which should make it possible for the `virt-launcher` container to boot the VMI without the need to handle `vfio-pci` driver binding (which requires privileged mode).

**Comment by anandgbhat** [comment]
@booxter Thanks for the detailed explanation. Will try sriov-network-operator.

**Comment by phoracek** [comment]
@anandgbhat I recently witnessed the same error. It happened when the SR-IOV driver was not set to `vfio-pci` but was kept on the default `netlink`. Could you please make sure you are using the right driver and let us know if it resolved the issue?
